sip/sip-004-materialized-view.md

```diff
@@ -0,0 +1,822 @@
+
+**Trie Structure**
+
+A trie is made up of nodes with radix 4, 16, 48, or 256, as well as leaves.  In
+the documentation below, these are called `node4`, `node16`, `node48`,
+`node256`, and `leaf` nodes.  An empty trie has a single `node256` as its root.
+Child pointers occupy one byte.
+
+**Notation**
+
+The notation `(ab)node256` means "a `node256` who descends from its parent via
+byte 0xab".
+
+The notation `node256[path=abcd]` means "a `node256` that has a shared prefix
+with is children `abcd`".
+
+**Lazy Expansion**
+
+If a leaf has a non-zero-byte path suffix, and another leaf is inserted that
+shares part of the suffix, the common bytes will be split off of the existing
+leaf to form a `node4`, whose two immediate children are the two leaves.  Each
+of the two leaves will store the path bytes that are unqiue to them.  For
+example, consider this trie with a root `node256` and a single leaf, located at
+path `aabbccddeeff00112233` and having value hash `123456`:
+
+```
+node256
+       \
+        (aa)leaf[path=bbccddeeff00112233]=123456
+```
+
+If the peer inserts the value hash `98765` at path `aabbccddeeff998877`, the
+single leaf's path will be split into a shared prefix and two distinct suffixes,
+as follows:
+
+```
+insert (aabbccddeeff998877, 98765)
+
+node256                            (00)leaf[path=112233]=123456
+       \                          /
+        (aa)node4[path-bbccddeeff]
+                                  \
+                                   (99)leaf[path=887766]=98765
+```
+
+Now, the trie encodes both `aabbccddeeff00112233=123456` and
+`aabbccddeeff99887766=98765`.
+
+**Node Promotion**
+
+As a node with a small radix gains children, it will eventually need to be
+promoted to a node with a higher radix.  A `node4` will become a `node16` when
+it receives its 5th child; a `node16` will become a `node48` when it receives
+its 17th child, and a `node48` will become a `node256` when it receives its 49th
+child.  A `node256` will never need to be promoted, because it has slots for
+child pointers with all possible byte values.
+
+For example, consider this trie with a `node4` and 4 children:
+
+```
+node256                                (00)leaf[path=112233]=123456
+       \                              /
+        \                            /  (01)leaf[path=445566]=67890
+         \                          /  /
+          (aa)node4[path=bbccddeeff]---
+                                    \  \
+                                     \  (02)leaf[path=778899]=abcdef
+                                      \
+                                       (99)leaf[path=887766]=98765
+```
+
+This trie encodes the following:
+   * `aabbccddeeff00112233=123456`
+   * `aabbccddeeff01445566=67890`
+   * `aabbccddeeff02778899=abcdef`
+   * `aabbccddeeff99887766=9876`
+
+Inserting one more node with a prefix `aabbccddeeff` will promote the
+intermediate `node4` to a `node16`:
+
+```
+insert (aabbccddeeff02aabbcc, 314159)
```

Comment:
Shouldn't this key `aabbccddeeff02aabbcc` result in a promotion of

`(02)leaf[path=778899]=abcdef`

to a `node2`, rather than promoting their parent?

Reply:
That's a typo.  I meant to type `insert (aabbccddeeff03aabbcc, 314159)`.  It doesn't make sense for there to be two children under the branch for `02`.

Reply:
Fixed (typo occurs elsewhere as well).